crates/cargo-contract/src/cmd/verify.rs

```diff
@@ -0,0 +1,236 @@
+    verbosity: VerbosityFlags,
+    /// Output the result in JSON format
+    #[clap(long, conflicts_with = "verbose")]
+    output_json: bool,
+}
+
+impl VerifyCommand {
+    pub fn run(&self) -> Result<VerificationResult> {
+        let manifest_path = ManifestPath::try_from(self.manifest_path.as_ref())?;
+        let verbosity: Verbosity = TryFrom::<&VerbosityFlags>::try_from(&self.verbosity)?;
+
+        // 1. Read the given metadata, and pull out the `BuildInfo`
+        let path = &self.contract;
+        let file = File::open(path)
+            .context(format!("Failed to open contract bundle {}", path.display()))?;
+
+        let metadata: ContractMetadata = serde_json::from_reader(&file).context(
+            format!("Failed to deserialize contract bundle {}", path.display()),
+        )?;
+        let build_info = if let Some(info) = metadata.source.build_info {
+            info
+        } else {
+            anyhow::bail!(
+                "\nThe metadata does not contain any build information which can be used to \
+                verify a contract."
+                .to_string()
+                .bright_yellow()
+            )
+        };
+
+        let build_info: BuildInfo =
+            serde_json::from_value(build_info.into()).context(format!(
+                "Failed to deserialize the build info from {}",
+                path.display()
+            ))?;
+
+        tracing::debug!(
+            "Parsed the following build info from the metadata: {:?}",
+            &build_info,
+        );
+
+        let build_mode = if metadata.image.is_some() {
+            BuildMode::Verifiable
+        } else {
+            build_info.build_mode
+        };
+
+        // 2. Check that the build info from the metadata matches our current setup.
+        // if the build mode is `Verifiable` we skip
+        if build_mode != BuildMode::Verifiable {
+            let expected_rust_toolchain = build_info.rust_toolchain;
+            let rust_toolchain = contract_build::util::rust_toolchain()
+                .expect("`rustc` always has a version associated with it.");
+
+            let rustc_matches = rust_toolchain == expected_rust_toolchain;
+            let mismatched_rustc = format!(
+            "\nYou are trying to `verify` a contract using the `{rust_toolchain}` toolchain.\n\
+             However, the original contract was built using `{expected_rust_toolchain}`. Please\n\
+             install the correct toolchain (`rustup install {expected_rust_toolchain}`) and\n\
+             re-run the `verify` command.",);
+            anyhow::ensure!(rustc_matches, mismatched_rustc.bright_yellow());
+
+            let expected_cargo_contract_version = build_info.cargo_contract_version;
+            let cargo_contract_version = semver::Version::parse(VERSION)?;
+
+            // Note, assuming both versions of `cargo-contract` were installed with the
+            // same lockfile (e.g `--locked`) then the versions of `wasm-opt`
```

Comment:
This may prove to be a common stumbling block, where the same version of `cargo-contract` is using a different `wasm-opt` version.

We could add the `wasm-opt` version to the `build_info`, or pin `wasm-opt` to the specific version to guarantee the version consistency even if the user does not install with `--locked`

Reply:
I think pinning one would be the best course of action here.